File: Go/Go_100-999/515.go
// Go
// leetcode 515
// https://leetcode.cn/problems/find-largest-value-in-each-tree-row/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func largestValues(root *TreeNode) (res []int) {
	if root == nil {
		return res
	}
	q := []*TreeNode{root}
	for len(q) != 0 {
		size, max := len(q), q[0].Val
		for i := 0; i < size; i++ {
			if q[i].Val > max {
				max = q[i].Val
			}
			if q[i].Left != nil {
				q = append(q, q[i].Left)
			}
			if q[i].Right != nil {
				q = append(q, q[i].Right)
			}
		}
		res = append(res, max)
		q = q[size : ]
	}
	return res
}

File: cpp/cpp_100-999/515.cc
// C++
// leetcode 515
// https://leetcode.cn/problems/find-largest-value-in-each-tree-row/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
    	vector<int> res;
    	if(root == nullptr)
    		return res;
    	queue<TreeNode*> q;
    	q.push(root);
    	int max_cur, size;
    	while(!q.empty()) {
    		size = q.size();
    		max_cur = INT_MIN;
    		while(size--) {
    			auto p = q.front();
    			q.pop();
    			max_cur = max(max_cur, p->val);
    			if(p->left) {
    				q.push(p->left);
    			}
    			if(p->right) {
    				q.push(p->right);
    			}
    		}
    		res.push_back(max_cur);
    	}
    	return res;
    }
};